include/oneapi/tbb/detail/_segment_table.h

```diff
@@ -260,34 +260,49 @@ class segment_table {
+            destroy_and_deallocate_table(current_segment_table, pointers_per_long_table);
             my_segment_table.store(my_embedded_table, std::memory_order_relaxed);
             zero_table(my_embedded_table, pointers_per_embedded_table);
         }
     }
 
     void extend_table_if_necessary(segment_table_type& table, size_type start_index, size_type end_index) {
-        // extend_segment_table if an active table is an embedded table
-        // and the requested index is not in the embedded table
+        // Extend segment table if an active table is an embedded one and the requested index is
+        // outside it
         if (table == my_embedded_table && end_index > embedded_table_size) {
             if (start_index <= embedded_table_size) {
+                // More than one thread can get here: the one that has assigned the first block and
+                // is in the process of allocating it now, and the one that saw the first block has
+                // been assigned already, but not yet allocated. This latter thread decides not to
+                // wait for the first one and extend the table itself.
                 try_call([&] {
-                    table = self()->allocate_long_table(my_embedded_table, start_index);
-                    // It is possible that the table was extended by the thread that allocated first_block.
-                    // In this case it is necessary to re-read the current table.
-
-                    if (table) {
-                        my_segment_table.store(table, std::memory_order_release);
-                    } else {
-                        table = my_segment_table.load(std::memory_order_acquire);
+                    segment_table_type new_table =
+                        self()->allocate_long_table(my_embedded_table, start_index);
+                    // It is possible that the table was extended by the thread that allocated first
+                    // block. In this case, the below CAS fails and re-reads the new table pointer.
+                    if (my_segment_table.compare_exchange_strong(
+                            table, new_table,
+                            /*memory order in case of a success*/std::memory_order_release,
+                            /*memory order in case of a failure*/std::memory_order_acquire))
+                    {
+                        // CAS was successful, update the local table pointer with now actual
+                        table = new_table;
+                    } else if (new_table) {
+                        // Other thread was the first to replace the segment table. Current thread's
+                        // table is not needed anymore, so destroying it.
+                        destroy_and_deallocate_table(new_table, pointers_per_long_table);
                     }
                 }).on_exception([&] {
                     my_segment_table_allocation_failed.store(true, std::memory_order_relaxed);
```

Comment:
Would this logic work correct while having one thread which allocates the long table and successfully stores it into the `my_segment_table` and another which allocates the long table and receives a `bad_alloc` while trying to allocate?
It seems like we need to double check this here.

Reply:
It depends on what logic we think is correct. Perhaps, not reporting allocation error in case it succeeded by some other thread makes sense, and double checking `my_segment_table` for updated state first can improve algorithm robustness in some cases. Although, I am not sure how frequent these cases can be. Also, it is not full cover anyway, as successful allocation can still sneak in after that double check. Full cover requires recording the failure state into single source of information such as `my_segment_table` or some kind of non-trivial synchronization between multiple information sources, but we don't have that kind of logic through the whole vector right now I guess. So, at most we can write something like the following instead:
```suggestion
                    // Last chance to overcome the failure, hoping that other thread has succeeded
                    // extending the table
                    table = get_table();
                    if (table == my_embedded_table) {
                        my_segment_table_allocation_failed.store(true, std::memory_order_relaxed);
                    }
```
Does it suffice though?

Reply:
I think we can keep it as-is for now since I believe this happens once in a blue moon and the ideal solution will require a lot of resources now.